### ISSUE TITLE
[BUG] Azure MySQL bind/create service key fails with broker timeout

### ISSUE THREAD

**Issue by claassen** [pivotal/cloud-service-broker#68]
## Description
Binding an Azure MySQL service to an app or trying to create a service key fails with a broker timeout error

## Expected Behavior
Service should successfully be bound to app.
Service key should be successfully created

## Actual Behavior
Both operations fail with error:
```
Unexpected Response
Response code: 504
CC code:       0
CC error code:
Request ID:    a2ebe399-7231-4588-7c62-4c6744c33229::94abf6d5-bbaf-4ead-9023-f166af8eb06e
Description:   {
  "description": "The request to the service broker timed out: https://cloud-service-broker.apps.use.sandbox.pcf.manulife.com/v2/service_instances/23139856-e296-44f0-a05c-c474221d3ba1/service_bindings/b9ae8436-a67a-4be1-8e0f-46407a36106f?accepts_incomplete=true",
  "error_code": "CF-HttpClientTimeout",
  "code": 10001,
  "http": {
    "uri": "https://cloud-service-broker.apps.use.sandbox.pcf.manulife.com/v2/service_instances/23139856-e296-44f0-a05c-c474221d3ba1/service_bindings/b9ae8436-a67a-4be1-8e0f-46407a36106f?accepts_incomplete=true",
    "method": "PUT"
  }
}
```

## Possible Fix
?

## Steps to Reproduce
1. Create service via `cf create-service csb-azure-mysql small csb-mysql-test-db`
2. Create service key via `cf create-service-key csb-mysql-test-db test-key`
3. Bind service to app via `cf bind-service test-app csb-mysql-test-db`

## Your Environment
Environment: Azure
Version: sb-0.1.0-rc.37-gcp-0.0.1-rc.76
Configuration:
```
csb-azure-mysql:
    provision:
      defaults: '{
       "skip_provider_registration": true,
       "resource_group":"mfc-use-pcf-psb"
      }'
```

**Comment by claassen** [comment]
Found this in broker logs:
```
{"timestamp":"1597848204.111438513","source":"cloud-service-broker","message":"cloud-service-broker.bind.unknown-error","log_level":2,"data":{"binding-id":"b9ae8436-a67a-4be1-8e0f-46407a36106f","error":"Error: Could not connect to server: Error 9000: Client with IP address '104.209.126.22' is not allowed to connect to this MySQL server. on brokertemplate/definition.tf line 42, in resource \"mysql_user\" \"newuser\": 42: resource \"mysql_user\" \"newuser\" { exit status 1","instance-id":"23139856-e296-44f0-a05c-c474221d3ba1","session":"169"}}
```

**Comment by erniebilling** [comment]
@claassen This error indicates that the network your broker is running on cannot reach the MySQL instance. Are your `cf push`ed applications (like the broker) using the same Azure credentials you provided to the broker to create Azure resources?

The default firewall rule for the Azure MySQL instance is 0.0.0.0-0.0.0.0 which should allow everything within the same tenant (or tenant and subscription) to access the instance.

If you are using separate networks, you can try the `authorized_network` parameter in your `cf create-service` command to set up a specific firewall rule to that network:

```
cf create-service csb-azure-mysql small csb-mysql-test-db -c `{"authorized_network":"<network name>"}'
```

**Comment by erniebilling** [comment]
@claassen Have you resolved this issue?

**Comment by claassen** [comment]
Still waiting on the required access to test this. Will follow up as soon as I can.

**Comment by claassen** [comment]
The authorized_network parameter seems to have resolved this issue, however I am now facing a different strange issue where it is complaining about permissions for provider registration even though I have skip_provider_registration set to true in my config.

Config:

```yaml
csb-azure-mysql:
    provision:
      defaults: '{
       "skip_provider_registration": true,
       "resource_group":"mfc-use-pcf-psb",
       "location", "eastus",
       "authorized_network": "/subscriptions/957108b5-005f-4b11-bae1-9e62ba39e857/resourceGroups/VNET_MFCv2-Internal-USE-Production-S5-PCFSandbox/providers/Microsoft.Network/virtualNetworks/MFCv2-Internal-USE-Production-S5-PCFSandbox/subnets/Servers"
      }'
```

Error when creating service instance:

```
Error: Error ensuring Resource Providers are registered.Terraform automatically attempts to register the Resource Providers it supports toensure it's able to provision resources.If you don't have permission to register Resource Providers you may wish to use
           the"skip_provider_registration" flag in the Provider block to disable this functionality.Please note that if you opt out of Resource Provider Registration and Terraform triesto provision a resource from a Resource Provider which is unregistered, then the
           errorsmay appear misleading - for example:> API version 2019-XX-XX was not found for Microsoft.FooCould indicate either that the Resource Provider "Microsoft.Foo" requires registration,but this could also indicate that this Azure Region doesn't support this API
           version.More information on the "skip_provider_registration" flag can be found here:https://www.terraform.io/docs/providers/azurerm/index.html#skip_provider_registrationOriginal Error: Cannot register provider Microsoft.DBforMariaDB with Azure Resource Manager:
           resources.ProvidersClient#Register: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailed" Message="The client 'b56542ac-6ea5-48d4-849a-4cf7a8285032' with object id
           'b56542ac-6ea5-48d4-849a-4cf7a8285032' does not have authorization to perform action 'Microsoft.DBforMariaDB/register/action' over scope '/subscriptions/957108b5-005f-4b11-bae1-9e62ba39e857' or the scope is invalid. If access was recently granted, please
           refresh your credentials.".  on brokertemplate/definition.tf line 33, in provider "azurerm":  33: provider "azurerm" { exit status 1
```

**Comment by claassen** [comment]
After some more experimentation it seems the "location" field may be causing this. I get the exact same error when I take out the "authorized_network":

```yaml
csb-azure-mysql:
    provision:
      defaults: '{
       "skip_provider_registration": true,
       "resource_group":"mfc-use-pcf-psb",
       "location", "eastus"
      }'
```

Error:

```
Error: Error ensuring Resource Providers are registered.Terraform automatically attempts to register the Resource Providers it supports toensure it's able to provision resources.If you don't have permission to register Resource Providers you may wish to use
           the"skip_provider_registration" flag in the Provider block to disable this functionality.Please note that if you opt out of Resource Provider Registration and Terraform triesto provision a resource from a Resource Provider which is unregistered, then the
           errorsmay appear misleading - for example:> API version 2019-XX-XX was not found for Microsoft.FooCould indicate either that the Resource Provider "Microsoft.Foo" requires registration,but this could also indicate that this Azure Region doesn't support this API
           version.More information on the "skip_provider_registration" flag can be found here:https://www.terraform.io/docs/providers/azurerm/index.html#skip_provider_registrationOriginal Error: Cannot register provider Microsoft.CustomProviders with Azure Resource
           Manager: resources.ProvidersClient#Register: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailed" Message="The client 'b56542ac-6ea5-48d4-849a-4cf7a8285032' with object
           id 'b56542ac-6ea5-48d4-849a-4cf7a8285032' does not have authorization to perform action 'Microsoft.CustomProviders/register/action' over scope '/subscriptions/957108b5-005f-4b11-bae1-9e62ba39e857' or the scope is invalid. If access was recently granted, please
           refresh your credentials.".  on brokertemplate/definition.tf line 33, in provider "azurerm":  33: provider "azurerm" { exit status 1
```

If I just remove the "location" field from the config it is successful.

**Comment by claassen** [comment]
Closing and re-opening as a new issue: https://github.com/pivotal/cloud-service-broker/issues/76
